### PR TITLE
Add mem buffer for kv event.

### LINKED ISSUE #271: implement a  limit memory buffer
https://github.com/pingcap/ticdc/blob/d2621b3f0f65f33567fa6bf772b93e8b2aee1128/cdc/puller/buffer.go#L40

currently, we will stop to consume kv events from tikv once the rest pipeline is slow.

implement a limit memory buffer and use it to buffer event from tikv, we should consume events from tikv ASAP  and fail the changfeed if the rest pipeline of feed is too slow.

### PR DESCRIPTION
**Pull request by july2993**
### What problem does this PR solve? 
fix #271 

### What is changed and how it works?
Add a limit memory buffer for kv event, consume the kv event
ASAP, once reach limited memory buffer will pull.Run will return error
and processor.Run will fail.

### Check List 

Tests 

 - Unit test
 - Manual test (add detailed scripts or steps below)
 temporarily change the default mem buffer value to a low value and will reach the `ErrReachLimit` case. 
```
107 [2020/02/24 19:20:29.829 +08:00] [INFO] [capture.go:117] ["stop to run processor"] ["changefeed id"=5ccf8990-2145-4baf-8286-2d07fb6d2de3] [error="reach limit"] [errorVerbose="reach limit\ngithub.com/pingcap/errors.AddStack\n\tgithub.com/pingcap/errors@v0.11.5-0.2    0190809092503-95897b64e011/errors.go:174\ngithub.com/pingcap/errors.Trace\n\tgithub.com/pingcap/errors@v0.11.5-0.20190809092503-95897b64e011/juju_adaptor.go:15\ngithub.com/pingcap/ticdc/cdc/puller.(*pullerImpl).Run.func2\n\tgithub.com/pingcap/ticdc@/cdc/puller/pu    ller.go:130\ngolang.org/x/sync/errgroup.(*Group).Go.func1\n\tgolang.org/x/sync@v0.0.0-20190911185100-cd5d95a43a6e/errgroup/errgroup.go:57\nruntime.goexit\n\truntime/asm_amd64.s:1357"]
108 [2020/02/24 19:20:30.407 +08:00] [DEBUG] [client.go:216] ["singleEventFeed quit"]
```

Code changes

Side effects

### CODE REVIEW

File: cdc/puller/buffer.go [hunk 1]
```diff
@@ -37,4 +55,112 @@ func (b Buffer) Get(ctx context.Context) (model.RegionFeedEvent, error) {
...
+var sizeOfVal = unsafe.Sizeof(model.RawKVEntry{})
+var sizeOfResolve = unsafe.Sizeof(model.ResolvedSpan{})
+
+func entrySize(e model.RegionFeedEvent) int {
+	if e.Val != nil {
+		return int(sizeOfVal) + len(e.Val.Key) + len(e.Val.Value)
+	} else if e.Resolved != nil {
+		return int(sizeOfResolve)
+	} else {
+		log.Fatal("unknow event type")
```
Comment (leoppro) [inline cdc/puller/buffer.go]: how about return `Sizeof(model.RegionFeedEvent)` here

File: cdc/puller/buffer.go [hunk 2]
```diff
@@ -37,4 +55,112 @@ func (b Buffer) Get(ctx context.Context) (model.RegionFeedEvent, error) {
...
+
+// AddEntry implements EventBuffer interface.
+func (b *memBuffer) AddEntry(ctx context.Context, entry model.RegionFeedEvent) error {
+	b.mu.Lock()
+	if b.limitter != nil && b.limitter.OverBucget() {
+		b.mu.Unlock()
+		return ErrReachLimit
+	}
+
+	b.mu.entries.PushBack(entry)
```
Comment (leoppro) [inline cdc/puller/buffer.go]: why we use `deque` but not `channel` here?

Comment (july2993) [reply cdc/puller/buffer.go]: `channel` is a bounded buffer

File: cdc/puller/buffer.go (hunk 1 above)
Comment (july2993) [reply cdc/puller/buffer.go]: this can't correctly get the size, we should know it and fix it if there' s new type.

File: cdc/puller/buffer.go [hunk 3]
```diff
@@ -37,4 +55,112 @@ func (b Buffer) Get(ctx context.Context) (model.RegionFeedEvent, error) {
...
+	} else {
+		log.Fatal("unknow event type")
+	}
+
+	return 0
+}
+
+// BlurResourceLimitter limit resource use.
+type BlurResourceLimitter struct {
+	bucget int64
```
Comment (amyangfei) [inline cdc/puller/buffer.go]: what does `bucget` mean, is it `budget` or `bucket`

Comment (july2993) [reply cdc/puller/buffer.go]: budget, fixed

**Pull request merged**